packages/styled-system/src/css.ts

```diff
@@ -7,40 +7,100 @@ import {
   objectToArrayNotation,
   runIfFn,
   merge,
+  isCustomBreakpoint,
 } from "@chakra-ui/utils"
 import { pseudoSelectors } from "./pseudo"
 import { parser } from "./parser"
 import { StyleObjectOrFn, CSSObject } from "./css.types"
 
-const defaultBreakpoints = ["40em", "52em", "64em"]
+interface Cache {
+  themeBreakpoints: string[]
+  breakpoints: string[]
+  breakpointValues: string[]
+  mediaQueries: (string | null)[]
+}
 
-function getBreakpoints(theme: any) {
-  const breakpoints = theme?.breakpoints ?? defaultBreakpoints
-  return breakpoints as string[]
+const cache: Cache = {
+  themeBreakpoints: [],
+  breakpoints: [],
+  breakpointValues: [],
+  mediaQueries: [],
 }
 
-const responsive = (styles: any) => (theme?: any) => {
-  const computedStyles: any = {}
-  const breakpoints = getBreakpoints(theme)
+interface BreakpointValueObj {
+  /**
+   * left side of a breakpoint object, the name, e.g. sm
+   */
+  breakpoints: string[]
+  /**
+   * right side of a breakpoint object, the size, e.g. 4
+   */
+  breakpointValues: string[]
+}
+
+/**
+ *
+ */
+const calculateBreakpointAndMediaQueries = (
+  themeBreakpoints: string[] = [],
+) => {
+  // caching here reduces execution time by factor 4-6x
+  const isCached = cache.themeBreakpoints === themeBreakpoints
```

Comment:
we might be able to remove caching here alltogether IF we never have a `ThemeProvider` with a different `theme` and thus potentially different `breakpoints`. Figured I'd leave it in as the original code was checking whether `theme.breakpoints` was defined at all. Unsure if this was due to a certain test scenario that I had to fix because of that, or because of some use case. @segunadebayo